### PR TITLE
[SPARK-23776][python][test] Check for needed components/files before running pyspark-sql tests

### PR DESCRIPTION
**Pull request by bersprockets**
## What changes were proposed in this pull request?

Change pyspark-sql tests to check the following:
- Spark was built with the Hive profile
- Spark scala tests were compiled

If either condition is not met, throw an exception with a message explaining how to appropriately build Spark.

These checks are similar to the ones found in the pyspark-streaming tests.

These required files will be missing if you follow the sbt build instructions. They are less likely to be missing if you follow the mvn build instructions (mvn compiles the test scala files, and there are mvn build instructions for running the pyspark tests).

## How was this patch tested?

For sbt build:
- run ./build/sbt package
- run python/run-tests --modules "pyspark-sql" --python-executables python2.7
- see failure, follow sbt instructions in exception message
- run test again
- see second failure (sbt only), follow sbt instructions in exception message
- run test again, verify success
- repeat for python3.4

For mvn build:
- run ./build/mvn -DskipTests clean package
- run python/run-tests --modules "pyspark-sql" --python-executables python2.7
- see failure, follow mvn instructions in exception message
- run test again, verify success
- repeat for python3.4

### CODE REVIEW

**Review by felixcheung — COMMENTED** [review]
only a very small number of tests require hive? 
https://github.com/bersprockets/spark/blob/8a965a51be6190f0db864ca7b1ba37269b3a55bc/python/pyspark/sql/tests.py#L3004

and for these it skip automatically (not fail) if jar is not built with hive. so I'm not sure we should raise exception here

**Comment by bersprockets** [comment]
Thanks @felixcheung . Turns out HiveSparkSubmitTests will fail if Spark is not built with the hive profile (AssertionError: 0 != 1).

In addition, at least one pyspark.sql.readwriter docstring test fails.

This PR uses  pyspark.sql.tests as the "leader of the pack" (run-tests.py gives it priortity 0 amongst the pyspark.sql tests) to check for prerequisites for its own tests as well as the sql docstring tests. The docstring tests can't make these checks.

I modeled this after pyspark/streaming/tests.py, which checks for prereqs and raises exceptions with a useful message so one can get past the error (although  pyspark/streaming/tests.py only checks for its own prereqs, not those required by streaming docstring tests).

**Comment by HyukjinKwon** [comment]
@bersprockets, do you have the error messages? I could (will) check it by myself in the following week but want to take a quick look if you already have them.

File: python/pyspark/sql/tests.py [hunk 1]
```diff
@@ -84,6 +85,31 @@
 from pyspark.sql.utils import AnalysisException, ParseException, IllegalArgumentException
 
 
+def found_file(pattern):
+    SPARK_HOME = os.environ["SPARK_HOME"]
+    files = glob.glob(os.path.join(SPARK_HOME, pattern))
+    return len(files) > 0
+
+
+def search_hive_assembly_jars():
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: Quick note: I think `check_hive_assembly_jars` would make more sense.

**Comment by felixcheung** [comment]
maybe the approach in hiveContextSQLTests is better for HiveSparkSubmitTests?
https://github.com/bersprockets/spark/blob/8a965a51be6190f0db864ca7b1ba37269b3a55bc/python/pyspark/sql/tests.py#L3112

**Comment by bersprockets** [comment]
@HyukjinKwon The HiveSparkSubmitTests error message is [here](https://gist.github.com/bersprockets/746ab6f2353bd5b54404d5cbb1df7403)

I propose the following:

- Fix HiveSparkSubmitTests according to @felixcheung's suggestion. After that fix, tests.py won't need the checks.
- Move the Hive assembly check to pyspark.sql.readwriter's _test() function.
- Move the test UDF check to pyspark.sql.udf's _test() function.

**Comment by HyukjinKwon** [comment]
> I modeled this after pyspark/streaming/tests.py, which checks for prereqs and raises exceptions with a useful message so one can get past the error (although pyspark/streaming/tests.py only checks for its own prereqs, not those required by streaming docstring tests).

I actually have been thinking about skipping and proceeding the tests in `pyspark/streaming/tests.py` with an explicit message as well. Can we skip and continue the tests? I think we basically should just skip the tests.

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -2977,6 +2977,20 @@ def test_create_dateframe_from_pandas_with_dst(self):
 
 class HiveSparkSubmitTests(SparkSubmitTests):
 
+    @classmethod
+    def setUpClass(cls):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: I think this way is more correct, as @felixcheung pointed out.

**Comment by bersprockets** [comment]
> actually have been thinking about skipping and proceeding the tests in pyspark/streaming/tests.py with an explicit message as well. Can we skip and continue the tests?

Hi @HyukjinKwon I just want to verify your comment: if hive assembly is missing, readwriter.py should not fail, but instead skip running its doctests. Also, in that case, there should be a message indicating that the tests were skipped.

**Comment by HyukjinKwon** [comment]
Yes, I feel sure that's more consistent and correct.

**Comment by HyukjinKwon** [comment]
@holdenk, I just saw https://issues.apache.org/jira/browse/SPARK-23853. I think this PR could fix it together if I understood correctly :-).

**Comment by bersprockets** [comment]
@HyukjinKwon That makes sense.

Note that when the tests are run using python/run-tests, run-tests.py steals stdout and stderr. I would need to make a small change to run-tests.py to detect when a tests is skipped (maybe through retcode) and print the message (from test test's stdout or stderr).

One other thing. I checked readwriter.py more closely, and there is only a single docstring test that requires Hive:

<pre>
>>> spark.read.table('tmpTable').dtypes
</pre>
I added <code># doctest: +SKIP</code> to that one line and all the tests passed. Rather than sometimes skipping all readerwriter tests, maybe we should just always skip that single test. 

udf.py, on the other hand, has lots of docstring tests that require the test udf files.

**Comment by bersprockets** [comment]
> can you investigate and try to explicitly skip some doctests conditionally?

@HyukjinKwon I will take a look to see how that can be done.

File: python/pyspark/sql/udf.py [hunk 3]
```diff
@@ -385,8 +385,23 @@ def registerJavaUDAF(self, name, javaClassName):
...
+    SPARK_HOME = os.environ["SPARK_HOME"]
+    filename_pattern = "sql/core/target/scala-*/test-classes/" + \
+                       "test/org/apache/spark/sql/JavaStringLength.class"
+    if not glob.glob(os.path.join(SPARK_HOME, filename_pattern)):
+        # if test udf files are not compiled, then skip the below doctests
+        # TODO: Need to communicate with outside world that these tests
+        # have been skipped.
+        m = pyspark.sql.udf
+        m.__dict__["UDFRegistration"].__dict__["registerJavaFunction"].__doc__ = ""
+        m.__dict__["UDFRegistration"].__dict__["registerJavaUDAF"].__doc__ = ""
```
Comment (HyukjinKwon) [inline python/pyspark/sql/udf.py]: ah.. hmm... yea this one was the last resort I was thinking ... let me investigate other possible ways for some more days.

**Comment by HyukjinKwon** [comment]
cc @viirya too.

File: python/pyspark/sql/readwriter.py [hunk 4]
```diff
@@ -964,6 +964,21 @@ def _test():
...
+    except TypeError:
+        hive_enabled = False
+
+    if not hive_enabled:
+        # if hive is not enabled, then skip doctests that need hive
+        # TODO: Need to communicate with outside world that this test
+        # has been skipped.
+        m = pyspark.sql.readwriter
+        m.__dict__["DataFrameReader"].__dict__["table"].__doc__ = ""
+
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/readwriter.py]: Thank you for pointer, @HyukjinKwon .

For `readwriter.py`, we had better test without Hive. How do you think , @HyukjinKwon and @bersprockets .
```
- spark = SparkSession.builder.enableHiveSupport().getOrCreate()
+ spark = SparkSession.builder.getOrCreate()
```

Comment (HyukjinKwon) [reply python/pyspark/sql/readwriter.py]: @bersprockets I was thinking like that but wanted to ask your thought per this PR. I am okay with either way.

Comment (bersprockets) [reply python/pyspark/sql/readwriter.py]: @HyukjinKwon @dongjoon-hyun I agree, it should go in PR #21141.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by bersprockets** [comment]
@HyukjinKwon This PR is mostly obsolete. I will close it and re-open something smaller... maybe a one-line documentation change to handle the missing UDF case for those who build with sbt.

**Pull request closed without merging**